Split core's PanicInfo and std's PanicInfo

m-ou-se opened the pull request:
`PanicInfo` is used in two ways:

1. As argument to the `#[panic_handler]` in `no_std` context.
2. As argument to the [panic hook](https://doc.rust-lang.org/stable/std/panic/fn.set_hook.html) in `std` context.

In situation 1, the `PanicInfo` always has a *message* (of type `fmt::Arguments`), but never a *payload* (of type `&dyn Any`).

In situation 2, the `PanicInfo` always has a *payload* (which is often a `String`), but not always a *message*.

Having these as the same type is annoying. It means we can't add `.message()` to the first one without also finding a way to properly support it on the second one. (Which is what https://github.com/rust-lang/rust/issues/66745 is blocked on.)

It also means that, because the implementation is in `core`, the implementation cannot make use of the `String` type (which doesn't exist in `core`): https://github.com/rust-lang/rust/blob/0692db1a9082380e027f354912229dfd6af37e78/library/core/src/panic/panic_info.rs#L171-L172

This also means that we cannot easily add a useful method like `PanicInfo::payload_as_str() -> Option<&str>` that works for both `&'static str` and `String` payloads.

I don't see any good reasons for these to be the same type, other than historical reasons.

---

This PR is makes 1 and 2 separate types. To try to avoid breaking existing code and reduce churn, the first one is still named `core::panic::PanicInfo`, and `std::panic::PanicInfo` is a new (deprecated) alias to `PanicHookInfo`. The crater run showed this as a viable option, since people write `core::` when defining a `#[panic_handler]` (because they're in `no_std`) and `std::` when writing a panic hook (since then they're definitely using `std`). On top of that, many definitions of a panic hook don't specify a type at all: they are written as a closure with an inferred argument type.

(Based on some thoughts I was having here: https://github.com/rust-lang/rust/pull/115561#issuecomment-1725830032)

---

For the release notes:

> We have renamed `std::panic::PanicInfo` to `std::panic::PanicHookInfo`. The old name will continue to work as an alias, but will result in a deprecation warning starting in Rust 1.82.0.
>
> `core::panic::PanicInfo` will remain unchanged, however, as this is now a *different type*.
> 
> The reason is that these types have different roles: `std::panic::PanicHookInfo` is the argument to the [panic hook](https://doc.rust-lang.org/stable/std/panic/fn.set_hook.html) in std context (where panics can have an arbitrary *payload*), while `core::panic::PanicInfo` is the argument to the [`#[panic_handler]`](https://doc.rust-lang.org/nomicon/panic-handler.html) in no_std context (where panics always carry a formatted *message*). Separating these types allows us to add more useful methods to these types, such as `std::panic::PanicHookInfo::payload_as_str()` and `core::panic::PanicInfo::message()`.

m-ou-se committed:
Split core's PanicInfo and std's PanicInfo.

m-ou-se commented:
@bors try

bors commented:
:hourglass: Trying commit 6707f29e814ce3a1278956afc7e5cfcb86c9e0a5 with merge 501befb16e66c3eabd7b0bef8ef0495d98d6410e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6707f29e814ce3a1278956afc7e5cfcb86c9e0a5","merge_sha":"501befb16e66c3eabd7b0bef8ef0495d98d6410e"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6238794561/job/16935382650) [(plain)](https://github.com/rust-lang-ci/rust/commit/501befb16e66c3eabd7b0bef8ef0495d98d6410e/checks/16935382650/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6238794561/job/16935728792)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/6238794561/job/16935728792","builder_name":"checks-actions"} -->

m-ou-se commented:
@bors try

bors commented:
:hourglass: Trying commit 6707f29e814ce3a1278956afc7e5cfcb86c9e0a5 with merge 47988eeabf6090a44f37845ce0a796693c491c97...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6707f29e814ce3a1278956afc7e5cfcb86c9e0a5","merge_sha":"47988eeabf6090a44f37845ce0a796693c491c97"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6238974857/job/16937997428)
Build commit: 47988eeabf6090a44f37845ce0a796693c491c97 (`47988eeabf6090a44f37845ce0a796693c491c97`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6238974857/job/16937997428"},"merge_sha":"47988eeabf6090a44f37845ce0a796693c491c97"} -->

m-ou-se commented:
@craterbot check

RalfJung commented:
I like this idea. :) I think it is somewhat similar to [my musings from a few years ago](https://github.com/rust-lang/rust/issues/66745#issuecomment-558558882), but goes a lot further: instead of just making these two cases the two variants of an enum, they become two entirely separate types. This means the parts they share have to be duplicated, but being able to have one of them in `std` does seem extremely useful.

RalfJung commented on library/std/src/panicking.rs:
```diff
@@ -725,20 +723,17 @@ fn rust_panic_with_hook(
+                    rtprintpanic!("aborting due to panic at {location}:\n{message}\n");
+                } else {
+                    rtprintpanic!("aborting due to panic at {location}\n");
+                }
```
Should this mention `always_abort`? Could be hard to figure out otherwise why is aborts.

m-ou-se replied:
`always_abort` isn't stable. The only stable way users can run into this error is through `Command::spawn` with `Command::pre_exec`. I don't think adding `always_abort` in the note would make it easier to trace it down to `Command`.

bjorn3 replied:
It would at least provide a clue to search for in the standard library source code.

m-ou-se replied:
> It would at least provide a clue to search for in the standard library source code.

If you're going to search through the stdlib code, then just searching for the message itself also works perfectly fine. ^^

I can't think of a real life scenario (in stable Rust) where mentioning the internal/unstable name `always_abort` makes things clearer rather than just a distraction. If there is a panic in a `pre_exec` hook, the problem is not that `always_abort` was called internally; instead, the problem is that there was a panic, so I believe it's better to focus on that.

bjorn3 replied:
> If you're going to search through the stdlib code, then just searching for the message itself also works perfectly fine. ^^

The current panic message is so generic I wouldn't be surprised if it would be used at multiple locations in the future. It also isn't useful if you already know about always_abort but don't recall the exact panic message for it. Even just mentioning that we are at a place where panics are not allowed would be an improvement.

m-ou-se replied:
I really don't think we should be showing unstable internal implementation details (such as always_abort) in messages that are shown to end users of Rust programs.

RalfJung commented on library/std/src/panicking.rs:
```diff
@@ -725,20 +723,17 @@ fn rust_panic_with_hook(
 
-    let mut info =
-        PanicInfo::internal_constructor(message, location, can_unwind, force_no_backtrace);
+    let mut info = PanicInfo::internal_constructor(location, can_unwind, force_no_backtrace);
```
```suggestion
    // We will set a payload before passing this to the hook.
    let mut info = PanicInfo::internal_constructor(location, can_unwind, force_no_backtrace);
```

m-ou-se replied:
I've removed the `set_payload` function and instead just set it all directly through the constructor, since there is no more need for a std PanicInfo to exist without a payload.

RalfJung commented on library/std/src/panicking.rs:
```diff
@@ -707,7 +705,7 @@ pub const fn begin_panic<M: Any + Send>(msg: M) -> ! {
 fn rust_panic_with_hook(
     payload: &mut dyn BoxMeUp,
-    message: Option<&fmt::Arguments<'_>>,
+    message: Option<fmt::Arguments<'_>>,
```
The message is now only used in case of must_abort, right? (And the payload is ignored in that case.) That seems worth mentioning in the doc comment.

m-ou-se replied:
I've implemented `Display` for the `PanicPayload`s, so it now uses that instead of a separate message argument.

m-ou-se commented:
Crater results:

15 failures are because of usage of the *unstable* `.message()` method (which changed signature). All of these crates explicitly opted out of stability using a nightly feature flag to use this unstable method.

Of those 15, two additionally also use `.payload()` on `core::panic::PanicInfo`, which no longer exists (in `core`) after this PR. We could add a `#[deprecated] fn payload` to core's PanicInfo that returns a dummy value (identical in behavior to before this PR), but these two crates would need changes anyway because of their unstable usage of `.message()`.

20 crates that failed to build because a dependency cannot be built.

7 unrelated failures (disk full, etc.).

And 2 regressions of stable code:

- `yacurses 0.2.4` imports `PanicInfo` from `core` rather than `std`, even though it is only used with a panic handler for `std::panic::set_hook`: https://github.com/Lokathor/yacurses/blob/8d6230a428f521f6de33a508e922c6da7aa8be6f/src/lib.rs#L52
- `damip/rswasm` (only on GitHub, not on crates.io) uses `core::panic::PanicInfo` in the panic hook signature, which is also only used with `std::panic::set_hook`: https://github.com/damip/rswasm/blob/5c58c9e079e5bdc3fef611072410ca4f4e6d99eb/src/lib.rs#L55

Neither of these two are cases where a panic handler/hook is a function that is used both as a no_std `#[panic_handler]` and as a std panic hook. In both cases, it is only used in combination with `std::panic::set_hook`, so they can be fixed by just changing `core` to `std`.

m-ou-se commented:
Looks like we might actually be able to do this.

m-ou-se committed:
Document difference between core and std's PanicInfo.

m-ou-se committed:
Add core::panic::PanicInfo::payload() for compatibility.

m-ou-se committed:
Fix doc link.

m-ou-se committed:
Update doc comment about core::panicking.

m-ou-se committed:
Remove core::panic::PanicInfo::internal_constructor.

It no longer needs to be public.

m-ou-se committed:
Remove std::panic::PanicInfo::internal_constructor+set_payload.

We can just set the payload immediately in the constructor,
and the constructor does not need to be public.

m-ou-se committed:
Mark some PanicInfo methods as #[inline] for consistency.

m-ou-se committed:
Move downcasting panic payload to str to a function.

m-ou-se committed:
Use unnamed lifetimes for [..]Payload impl blocks.

m-ou-se committed:
Impl Display for PanicPayload to simplify things.

m-ou-se committed:
Reorder body of begin_panic for consistency.

In the other functions, we put the struct and impl blocks first,
such that the return expression can be at the end of the body as usual.

m-ou-se commented:
@rfcbot merge

This splits `core::panic::PanicInfo` and `std::panic::PanicInfo` into two separate types. See the [intial comment](https://github.com/rust-lang/rust/pull/115974#issue-1903329164) for details and motivation.

As shown by the [crater results](https://github.com/rust-lang/rust/pull/115974#issuecomment-1732280770), almost nothing breaks, since these two paths are normally not mixed.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6380122783/job/17313931946) [(plain)](https://github.com/rust-lang/rust/commit/c59dc2b2f8dd9148be218b015ff11d6c7d39c12d/checks/17313931946/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error: unclosed HTML tag `dyn`
 --> library/core/src/panicking.rs:5:10
  |
5 | //! a Box<dyn Any> containing any type of value. Because of this, std::panic::PanicInfo is a
  |
  = note: `-D rustdoc::invalid-html-tags` implied by `-D warnings`

error: could not document `core`
error: could not document `core`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name core library/core/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/doc -Zunstable-options --check-cfg 'names()' --check-cfg 'values()' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=6f6461ce7e567ed5 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/release/deps --cfg=bootstrap -Csymbol-mangling-version=legacy -Zunstable-options -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(stdarch_intel_sde)' '--check-cfg=values(no_fp_fmt_parse)' '--check-cfg=values(no_global_oom_handling)' '--check-cfg=values(no_rc)' '--check-cfg=values(no_sync)' '--check-cfg=values(freebsd12)' '--check-cfg=values(freebsd13)' '--check-cfg=values(backtrace_in_libstd)' '--check-cfg=values(target_vendor,"unikraft")' '--check-cfg=values(target_env,"libnx")' '--check-cfg=values(target_os,"teeos","hurd")' '--check-cfg=values(target_arch,"asmjs","spirv","nvptx","xtensa","mips32r6","mips64r6","csky")' --document-private-items --document-hidden-items -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.75.0-nightly (e1ca95547 2023-10-02)' '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")' '-Zcrate-attr=warn(rust_2018_idioms)' -Z unstable-options --resource-suffix 1.75.0 --markdown-css rust.css --markdown-no-toc --index-page /checkout/src/doc/index.md` (exit status: 1)
Build completed unsuccessfully in 0:00:08
  local time: Mon Oct  2 12:46:52 UTC 2023
  network time: Mon, 02 Oct 2023 12:46:52 GMT
##[error]Process completed with exit code 1.
```

</details>

Amanieu commented:
This may become a problem if, in the future, we want to unify std and core. Currently std is a strict superset of the API provided by core, but this PR makes the same path refer to a different type depending on whether it was imported through std or core.

@rfcbot concern std/core unification

m-ou-se committed:
Fix invalid markdown/html.

m-ou-se commented:
> This may become a problem if, in the future, we want to unify std and core.

I don't think fully unifying core and std makes sense. Regardless, we will always allow for Rust programs to exist in a world where alloc and the non-core parts of std don't exist. (Whether through a core/std split, or through cfg(feature) in std, etc.) In such (no-std) code, the concept of a `String` message or a `Box<dyn Any>` payload is meaningless.

Using the same argument type for `std::panic::set_hook` and `#[panic_handler]` was a mistake, and this seems like the most backwards-compatible way forward.

> Currently std is a strict superset of the API provided by core, but this PR makes the same path refer to a different type depending on whether it was imported through std or core.

(The public API of core is technically already not a subset of the public api of std. core::panic!() and std::panic!() are also subtly different.)

core::panic::PanicInfo not being available in std seems fine to me, because it is useless when using std: std already defines its own #[panic_handler], so you cannot define your own, which would be the only reason you'd need core's PanicInfo.

m-ou-se commented on library/core/src/panic/panic_info.rs:
```diff
@@ -128,6 +75,24 @@ impl<'a> PanicInfo<'a> {
 
+    /// Returns the payload associated with the panic.
+    ///
+    /// On `core::panic::PanicInfo`, this method never returns anything useful.
+    /// It only exists because of compatibility with [`std::panic::PanicInfo`],
+    /// which used to be the same type.
```
Hm, I don't think that's the right way to handle such panics.

I'll post my thoughts in a comment on the ACP.

workingjubilee commented:
Even if we unified `core` and `std` so that they compiled as a single unit, in what world would we suddenly make it impossible to write `core::module::path::to::function` and `std::module::path::to::function`? That would **break back compat**. I habitually write `core` when writing against APIs that I know to live in `core`.

Amanieu commented:
> Even if we unified `core` and `std` so that they compiled as a single unit, in what world would we suddenly make it impossible to write `core::module::path::to::function` and `std::module::path::to::function`? That would **break back compat**. I habitually write `core::module::path::to::function` when writing against APIs that I know to live in `core`.

The idea is that once the crates are unified into a single `std` crate, `core` can be turned into a `use std as core` in the prelude.

workingjubilee commented:
I know it's less "pretty", but that still leaves the possibility open of
```rust
pub mod core {
    pub use std::*;
    pub use ::crate::panic_impl::CorePanicInfo as PanicInfo;
}
```

Or something like that.

BurntSushi commented:
It definitely bothers me that `core::path::to::foo` and `std::path::to::foo` do not refer to the same thing.

> (The public API of core is technically already not a subset of the public api of std. core::panic!() and std::panic!() are also subtly different.)

I was unaware of this. Still, it's not clear to me that this means we should make it worse.

Is there a path we can take here via an edition boundary where a panic handler uses a completely new type that is present in both `core` and `std`? Would that help?

workingjubilee commented:
Isn't that what this PR does, effectively?

BurntSushi commented:
@workingjubilee My understanding is that this PR makes `core::panic::PanicInfo` and `std::panic::PanicInfo` refer to different types. What I'm thinking is some shenanigans with an edition boundary to make panic handlers use a type other than `core::panic::PanicInfo`. Say, `core::panic::PanicHandlerInfo`. Then perhaps `core::panic::PanicInfo` and `std::panic::PanicInfo` could refer to the same type and be used exclusively for hooks.

(I am fairly out of my depth here in multiple ways, so much so that I might be making categorical errors. But the gist of my suggestion is 1) edition boundary, 2) distinct types using different names entirely, 3) `core::foo::bar` and `std::foo::bar` should refer to the same thing.)

workingjubilee commented:
Sooo, what happens when I'm on Rust 2024 but I use a crate with a 2021 `#[panic_handler]`?

workingjubilee commented:
I suppose we can introduce a new type, `impl AsRef<NewPanicInfo> for PanicInfo`, and have `#[panic_handler]`, as it is an attribute and thus allowed to be a proc macro (basically), unconditionally elaborate `&PanicInfo` to `&NewPanicInfo`. That would be _less_ complicated, probably.

bors commented:
:umbrella: The latest upstream changes (presumably #110303) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

m-ou-se commented:
> It definitely bothers me that `core::path::to::foo` and `std::path::to::foo` do not refer to the same thing.

@BurntSushi How would you feel about instead naming the two types:

- `core::panic::PanicHandlerInfo`, and
- `std::panic::PanicHookInfo`,

and then having two _deprecated aliases_:

- `#[deprecated] pub type PanicInfo = PanicHandlerInfo;` in `core::panic`, and
- `#[deprecated] pub type PanicInfo = PanicHookInfo;` in `std::panic`.

Then the types will have different names, and the old `{core,std}::panic::PanicInfo` name will still work and resolve differently in std and core, but using that name will result in a deprecation warning, such that it will not be a silent surprise that `core::panic::PanicInfo` and `std::panic::PanicInfo` are different types.

m-ou-se committed:
Formatting.

m-ou-se committed:
Rename std::panic::PanicInfo to PanicHookInfo.

m-ou-se committed:
Fix deprecation version.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7724041960/job/21055357572) [(plain)](https://github.com/rust-lang/rust/commit/0393ef52cbf2325aa51af70b2f8465351650b590/checks/21055357572/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#12 writing image sha256:45b59d8aecf207392344742e0dd02e51a4c96c9450a21f79900a3dfc0943c724 done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Jan 31 09:43:11 UTC 2024
  network time: Wed, 31 Jan 2024 09:43:11 GMT
  network time: Wed, 31 Jan 2024 09:43:11 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
##[endgroup]
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished release [optimized] target(s) in 1.35s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --no-default-features --mini-tests --std-tests`
Using system GCC
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
   Compiling profile v0.0.0 (/checkout/src/tools/rust-analyzer/crates/profile)
   Compiling rowan v0.15.15
   Compiling stdx v0.0.0 (/checkout/src/tools/rust-analyzer/crates/stdx)
   Compiling text-edit v0.0.0 (/checkout/src/tools/rust-analyzer/crates/text-edit)
error: use of deprecated type alias `std::panic::PanicInfo`: use `PanicHookInfo` instead
  --> crates/stdx/src/panic_context.rs:23:46
   |
23 |         let hook = move |panic_info: &panic::PanicInfo<'_>| {
   |
   = note: `-D deprecated` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(deprecated)]`

```

</details>

lnicola commented:
`rust-analyzer` builds on stable. Can we `#[allow]` the deprecated type until we bump our MSRV to a version that has `PanicHookInfo`?

m-ou-se committed:
Move deprecation of std::panic::PanicInfo to 1.80.0.

m-ou-se commented:
> This may become a problem if, in the future, we want to unify std and core. Currently std is a strict superset of the API provided by core, but this PR makes the same path refer to a different type depending on whether it was imported through std or core.

I made std::panic::PanicInfo a (deprecated) alias to the newly named PanicHookInfo. If we ever want to fully unify core+std, we'd have to remove that alias or make it conditional on an edition, which is possible but not something we have to do right now.

@Amanieu Does that address your concern?

BurntSushi commented:
I'm still uneasy about, but I feel like the deprecation mitigation probably makes this change worth it on balance.

bors commented:
:umbrella: The latest upstream changes (presumably #122218) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Amanieu commented on library/core/src/error.md:
```diff
@@ -17,7 +17,7 @@ The following are the primary interfaces of the panic system and the
 
 * [`panic!`] and [`panic_any`] (Constructing, Propagated automatically)
-* [`PanicInfo`] (Reporting)
+* [`PanicHookInfo`] (Reporting)
```
Should this list mention `PanicInfo` as well? This is in the core docs after all.

m-ou-se commented on library/std/src/panic.rs:
```diff
@@ -4,11 +4,167 @@
+    /// use std::panic;
+    ///
+    /// panic::set_hook(Box::new(|panic_info| {
+    ///     if let Some(s) = panic_info.payload().downcast_ref::<&str>() {
```
> Actually, we should provide a method specifically to extract a string from the payload since that is what most users will want to do.

Yes! That's one of the reasons I want to split these types, so we can add that method.

m-ou-se replied:
I've updated the example, and also added PanicHookInfo::payload_as_str() as unstable.

m-ou-se commented on library/std/src/panic.rs:
```diff
@@ -4,11 +4,167 @@
+pub struct PanicHookInfo<'a> {
+    payload: &'a (dyn Any + Send),
+    location: &'a Location<'a>,
+    can_unwind: bool,
```
I assumed it's still useful for the hook to know whether unwinding may follow or not. (Through `can_unwind()`.) It's not something I use though. Happy to remove it.

Amanieu replied:
This doesn't give you the information you want though, it's still `true` if the code is compiled with panic=abort. It's only false for panics from nounwind functions. I'd rather just remove it here.

m-ou-se replied:
It is currently used in two places in rust-lang/rust:

https://github.com/rust-lang/rust/blob/b71e8cbaf2c7cae4d36898fff1d0ba19d9233082/library/proc_macro/src/bridge/client.rs#L286-L292

and

https://github.com/rust-lang/rust/blob/b71e8cbaf2c7cae4d36898fff1d0ba19d9233082/library/test/src/lib.rs#L120-L139

m-ou-se replied:
So I think we should leave it in place for now (as unstable). And perhaps replace it in the future by a `will_unwind()` or something.

m-ou-se committed:
Downcast panic payload to String too in example.

m-ou-se committed:
Add note on panic payload type.

m-ou-se committed:
Mention core's PanicInfo in error.md.

m-ou-se committed:
Fix display of panic message in recursive panic.

m-ou-se committed:
Add PanicHookInfo::payload_as_str().

m-ou-se committed:
Bump deprecation of std's PanicInfo alias to 1.82.0.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9111205473/job/25047862266) [(plain)](https://github.com/rust-lang/rust/commit/feccb27d18a4e7c7ddc252820965f9b1e205abdb/checks/25047862266/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
7 |     if let Some(s) = panic_info.payload_as_str() {
  |                                 ^^^^^^^^^^^^^^
  |
  = note: see issue #125175 <https://github.com/rust-lang/rust/issues/125175> for more information
  = help: add `#![feature(panic_payload_as_str)]` to the crate attributes to enable

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0658`.
For more information about this error, try `rustc --explain E0658`.
Couldn't compile the test.

failures:
    library/std/src/panic.rs - panic::PanicHookInfo<'a>::payload_as_str (line 112)

test result: FAILED. 1151 passed; 1 failed; 20 ignored; 0 measured; 0 filtered out; finished in 32.19s

error: doctest failed, to rerun pass `-p std --doc`
  local time: Thu May 16 11:28:11 UTC 2024
  network time: Thu, 16 May 2024 11:28:12 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 188b2aba46b7b294ce469c8ef571e83201750630 has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu 188b2aba46b7b294ce469c8ef571e83201750630 -->
<!-- homu: {"type":"Approved","sha":"188b2aba46b7b294ce469c8ef571e83201750630","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 188b2aba46b7b294ce469c8ef571e83201750630 with merge b79bc519ba8f1ddd629038ce8e9903aa013fe67b...
<!-- homu: {"type":"BuildStarted","head_sha":"188b2aba46b7b294ce469c8ef571e83201750630","merge_sha":"b79bc519ba8f1ddd629038ce8e9903aa013fe67b"} -->

rust-log-analyzer commented:
The job **`aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9429013472/job/25974954881) [(plain)](https://github.com/rust-lang-ci/rust/commit/b79bc519ba8f1ddd629038ce8e9903aa013fe67b/checks/25974954881/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING] core::build_steps::tool::ToolBuild { compiler: Compiler { stage: 0, host: aarch64-apple-darwin }, target: aarch64-apple-darwin, tool: "tidy", path: "src/tools/tidy", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "" } -- 20.446
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: aarch64-apple-darwin }, target: aarch64-apple-darwin } -- 0.000
fmt check
fmt: skipped untracked file AWSCLIV2.pkg
##[error]Diff in /Users/runner/work/rust/rust/library/std/src/panicking.rs at line 768:
                 // recursive panics. However if the message is just a string, no user-defined
                 // code is involved in printing it, so that is risk-free.
                 let message: &str = payload.as_str().unwrap_or_default();
-                rtprintpanic!("panicked at {location}:\n{message}\nthread panicked while processing panic. aborting.\n");
+                rtprintpanic!(
+                    "panicked at {location}:\n{message}\nthread panicked while processing panic. aborting.\n"
             }
             panic_count::MustAbort::AlwaysAbort => {
                 // Unfortunately, this does not print a backtrace, because creating
                 // Unfortunately, this does not print a backtrace, because creating
fmt error: Running `"/Users/runner/work/rust/rust/build/aarch64-apple-darwin/rustfmt/bin/rustfmt" "--config-path" "/Users/runner/work/rust/rust" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/Users/runner/work/rust/rust/compiler/rustc_incremental/src/persist/mod.rs" "/Users/runner/work/rust/rust/compiler/rustc_incremental/src/persist/dirty_clean.rs" "/Users/runner/work/rust/rust/compiler/rustc_incremental/src/persist/save.rs" "/Users/runner/work/rust/rust/compiler/rustc_incremental/src/assert_dep_graph.rs" "/Users/runner/work/rust/rust/compiler/rustc_incremental/src/errors.rs" "/Users/runner/work/rust/rust/compiler/rustc_incremental/src/lib.rs" "/Users/runner/work/rust/rust/compiler/rustc_type_ir_macros/src/lib.rs" "/Users/runner/work/rust/rust/compiler/rustc_hir_pretty/src/lib.rs" "/Users/runner/work/rust/rust/library/std/src/io/buffered/tests.rs" "/Users/runner/work/rust/rust/library/std/src/io/buffered/bufreader/buffer.rs" "/Users/runner/work/rust/rust/library/std/src/io/buffered/mod.rs" "/Users/runner/work/rust/rust/library/std/src/io/buffered/linewriter.rs" "/Users/runner/work/rust/rust/library/std/src/io/buffered/linewritershim.rs" "/Users/runner/work/rust/rust/library/std/src/io/buffered/bufreader.rs" "/Users/runner/work/rust/rust/library/std/src/io/buffered/bufwriter.rs" "/Users/runner/work/rust/rust/library/std/src/io/error/tests.rs" "/Users/runner/work/rust/rust/library/std/src/io/error/repr_bitpacked.rs" "/Users/runner/work/rust/rust/library/std/src/io/error/repr_unpacked.rs" "/Users/runner/work/rust/rust/library/std/src/io/prelude.rs" "/Users/runner/work/rust/rust/library/std/src/io/copy/tests.rs" "/Users/runner/work/rust/rust/library/std/src/io/tests.rs" "/Users/runner/work/rust/rust/library/std/src/io/stdio.rs" "/Users/runner/work/rust/rust/library/std/src/io/cursor.rs" "/Users/runner/work/rust/rust/library/std/src/io/stdio/tests.rs" "/Users/runner/work/rust/rust/library/std/src/io/impls.rs" "/Users/runner/work/rust/rust/library/std/src/io/mod.rs" "/Users/runner/work/rust/rust/library/std/src/io/impls/tests.rs" "/Users/runner/work/rust/rust/library/std/src/io/error.rs" "/Users/runner/work/rust/rust/library/std/src/io/cursor/tests.rs" "/Users/runner/work/rust/rust/library/std/src/io/util/tests.rs" "/Users/runner/work/rust/rust/library/std/src/io/util.rs" "/Users/runner/work/rust/rust/library/std/src/io/copy.rs" "/Users/runner/work/rust/rust/library/std/src/lib.rs" "/Users/runner/work/rust/rust/library/std/src/path/tests.rs" "/Users/runner/work/rust/rust/library/std/src/backtrace/tests.rs" "/Users/runner/work/rust/rust/library/std/src/env.rs" "/Users/runner/work/rust/rust/library/std/src/error.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/rustc_smir/convert/mir.rs" "/Users/runner/work/rust/rust/library/std/src/hash/mod.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/rustc_smir/convert/abi.rs" "/Users/runner/work/rust/rust/library/std/src/hash/random.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/rustc_smir/convert/ty.rs" "/Users/runner/work/rust/rust/library/std/src/panicking.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/rustc_smir/convert/mod.rs" "/Users/runner/work/rust/rust/library/std/src/rt.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/rustc_smir/convert/error.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/rustc_smir/context.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/rustc_smir/builder.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/rustc_smir/alloc.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/rustc_smir/mod.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/lib.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/rustc_internal/pretty.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/rustc_internal/mod.rs" "/Users/runner/work/rust/rust/compiler/rustc_smir/src/rustc_internal/internal.rs" "/Users/runner/work/rust/rust/library/std/src/sys_common/wtf8/tests.rs" "/Users/runner/work/rust/rust/library/std/src/sys_common/process.rs" "/Users/runner/work/rust/rust/library/std/src/sys_common/fs.rs" "/Users/runner/work/rust/rust/library/std/src/sys_common/tests.rs" "/Users/runner/work/rust/rust/library/std/src/sys_common/backtrace.rs" "/Users/runner/work/rust/rust/library/std/src/sys_common/mod.rs" "/Users/runner/work/rust/rust/library/std/src/sys_common/thread_local_dtor.rs" "/Users/runner/work/rust/rust/library/std/src/sys_common/lazy_box.rs" "/Users/runner/work/rust/rust/compiler/rustc/src/main.rs" "/Users/runner/work/rust/rust/compiler/rustc_incremental/src/persist/data.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Sat Jun  8 13:15:13 UTC 2024
  network time: Sat, 08 Jun 2024 13:15:13 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9429013472/job/25975004111)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9429013472/job/25975004111","builder_name":"checks-actions"} -->

m-ou-se committed:
Formatting.

m-ou-se commented:
@bors r=Amanieu

bors commented:
:pushpin: Commit a6e23b126b925b270663436a1d2f4d24cc5f4711 has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu a6e23b126b925b270663436a1d2f4d24cc5f4711 -->
<!-- homu: {"type":"Approved","sha":"a6e23b126b925b270663436a1d2f4d24cc5f4711","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

RalfJung commented on library/core/src/panic/panic_info.rs:
```diff
@@ -1,98 +1,40 @@
 
     /// If the `panic!` macro from the `core` crate (not from `std`)
     /// was used with a formatting string and some additional arguments,
     /// returns that message ready to be used for example with [`fmt::write`]
```
The doc comment can probably be updated now, it still indicates that this may sometimes not be available.

RalfJung commented on library/core/src/panic/panic_info.rs:
```diff
@@ -128,6 +70,24 @@ impl<'a> PanicInfo<'a> {
 
+    /// Returns the payload associated with the panic.
+    ///
+    /// On `core::panic::PanicInfo`, this method never returns anything useful.
```
This is a bit confusing since `Self` is `core::panic::PanicInfo`, but here it sounds like you are talking about a different type.

RalfJung commented on library/core/src/panic/panic_info.rs:
```diff
@@ -128,6 +70,24 @@ impl<'a> PanicInfo<'a> {
+    ///
+    /// [`std::panic::PanicHookInfo`]: ../../std/panic/struct.PanicHookInfo.html
+    /// [`std::panic::PanicHookInfo::payload`]: ../../std/panic/struct.PanicHookInfo.html#method.payload
+    #[deprecated(since = "1.77.0", note = "this never returns anything useful")]
```
The "since" here seems wrong.
There's some sort of magic string that can be put here to become the current version on the next bootstrap bump, but I can never remember the exact string...

m-ou-se replied:
Looks like that magic string doesn't work in `#[deprecated]`, only in `#[stable]`. I'll update it to the correct version (1.81.0).

RalfJung commented on library/std/src/panic.rs:
```diff
@@ -4,11 +4,215 @@
+            formatter.write_str(payload)?;
+        } else if let Some(payload) = self.payload.downcast_ref::<String>() {
+            formatter.write_str(":\n")?;
+            formatter.write_str(payload)?;
```
Can't this use `payload_as_str`?

RalfJung commented on library/std/src/panicking.rs:
```diff
@@ -248,13 +248,7 @@ fn default_hook(info: &PanicInfo<'_>) {
-            None => "Box<dyn Any>",
-        },
-    };
+    let msg = payload_as_str(info.payload());
```
Can't this use `info.payload_as_str()`?

m-ou-se replied:
I want to use the same fallback string (`"Box<dyn Any>"`) as in the other cases here, so I re-use the same helper function here. And that helper function doesn't use PanicHookInfo::payload_as_str(), because it is also used in places where there is no PanicHookInfo.

RalfJung replied:
Makes sense... looking at my own code doing these downcasts, it's mostly working in the return value from `catch_unwind`, where there's also no `PanicHookInfo`. Maybe we should have some more widely used helper method that works on all `Box<dyn Any + ...>` or `&dyn Any` or so.

m-ou-se commented:
@RalfJung Thanks for catching those things! I've addressed your comments in https://github.com/rust-lang/rust/pull/126322

Manishearth commented:
This caused a minor regression in nightly: https://github.com/rust-lang/rust/issues/126766
